### ISSUE TITLE
How to Name Handlers

### ISSUE THREAD

**Issue by freeformflow** [pandastrike/panda-sky#70]
Right now, we rely on a convention for Lambda names:
```
environment-appName-resource-HttpVerb
```

Sky assumes this pattern when it creates and names Lambdas, but that's not enforced in the actual code that constitutes Lambdas.  So, within the code, we are relying on convention.  

We should solidify our design and then enforce it, both in the CloudFormation template and the Lambda code.

**Comment by automatthew** [comment]
With the request Mapping Template now transforming the Lambda events into Panda Request Representation, we now have access to the HTTP Method in the `event` object passed to the Lambda handler.

This opens up some new possibilities that I need to muse on.

**Comment by automatthew** [comment]
Consider this as a new approach to the Sky Lambda dispatcher:

```coffee
exports.handle = (request, context, callback) ->
  resource = Blurb.getResource(context)
  methodName = request.method
  resource[methodName](request, context)
  .then (result) ->
    callback null, result
  .catch (e) ->
    callback e
```

**Comment by automatthew** [comment]
`resource = Blurb.getResource(context)`  could be handled in at least a couple of ways.

1. using conventions in `context.functionName` as at present
2. determining the API Gateway "resource path" using a Mapping Template and `$context.resource_path`.  This provides the parameterized resource path, e.g. "/request-mapping-test/{key}"

**Comment by automatthew** [comment]
Currently in the CloudFormation template:

```yaml
  BlurbGetLambdaHandler:
    Type: 'AWS::Lambda::Function'
    Properties:
      Description: Handler for APIblurb9
      FunctionName: blurb9-staging-blurb-get
```

**Comment by automatthew** [comment]
We could simplify by having one Lambda per resource, instead of per (resource X method).

**Comment by automatthew** [comment]
In API Gateway, we can determine the deployment environment using Mapping Template's `$context.stage`.  This means we don't necessarily need to put it in the Lambda FunctionName, just to get Sky dispatching working.  We may still need it as part of namespacing Lambdas themselves.

**Comment by automatthew** [comment]
https://docs.aws.amazon.com/lambda/latest/dg/versioning-aliases.html

**Comment by automatthew** [comment]
https://aws.amazon.com/blogs/compute/using-api-gateway-stage-variables-to-manage-lambda-functions/

**Comment by automatthew** [comment]
Implication of the above two links is that we may be able use API Gateway stageVariables instead of putting stage names in our function names.

**Comment by automatthew** [comment]
semi-helpful discussion:  https://forums.aws.amazon.com/thread.jspa?messageID=681736

**Comment by automatthew** [comment]
I have not found any AWS native way to deal with the application namespacing (e.g. "blurb9") though.

**Comment by automatthew** [comment]
It's risky to jump straight into trusting all the AWS affordances I've just described.  Here's an idea for the next safe step:

```coffee
exports.handle = (request, context, callback) ->
  [_app, _stage, resource] = context.functionName.split('-')
  resource = Blurb9.getResource(resource)
  methodName = request.method
  resource[methodName](request, context)
  .then (result) ->
    callback null, result
  .catch (e) ->
    callback e
```

**Comment by automatthew** [comment]
This moves us one---very important---step closer to isomorphism between the AWS concepts and the Sky API concepts.

**Comment by automatthew** [comment]
```coffee
Blurb9 = Sky.API
  resources:
    Blurbs: require "./blurbs"
    Blurb: require "./blurb"
```

**Comment by dyoder** [comment]
- One lambda per resources makes sense, given that we're providing helper methods to unpack this anyway.

- We can also make the callback go away with a helper. 

- Why is the request not part of the context?

**Comment by automatthew** [comment]
> We can also make the callback go away with a helper.

Yeah, in a variety of ways.  My code example is just a riff on the current helper.

> Why is the request not part of the context?

Because AWS has them as separate (confusing) arguments.  In these examples, I'm trying to cleave close to our existing work and the AWS apis, to make it easier to see what's going on.

**Comment by automatthew** [comment]
Both "API Gateway Mapping Templates" and "Lambda handler functions" use the term `context`, and it means totally different things.  Which is unfortunate and confusing, but eh.

**Comment by automatthew** [comment]
With the `sky.method` wrapper I describe elsewhere, we get to provide our own interface to these resource-method functions, so we can easily reorganize the information into a single `skyContext` if we so desire.
